src/Tools/dotnet-monitor/HostBuilder/HostBuilderHelper.cs

```diff
@@ -133,12 +150,38 @@ public static IHostBuilder CreateHostBuilder(HostBuilderSettings settings)
                             urls,
                             metricsOptions.Enabled.GetValueOrDefault(MetricsOptionsDefaults.Enabled) ? metricUrls : Array.Empty<string>(),
                             settings.Authentication.KeyAuthenticationMode != KeyAuthenticationMode.NoAuth);
-
-                        // Since the endpoints have already been defined on the KestrelServerOptions, clear out the urls
-                        // so that the address binder does not log a warning about overriding urls as specified by configuration.
-                        context.Configuration[WebHostDefaults.ServerUrlsKey] = string.Empty;
                     })
                     .UseStartup<Startup>();
+                })
+                .ConfigureHostConfiguration((IConfigurationBuilder builder) =>
+                {
+                    // Restore the Urls option to the configuration provider that originally provided the value
+                    // before it was cleared during the IWebHostBuilder configuration callback.
+                    if (!string.IsNullOrEmpty(aspnetUrls) &&
+                        builder.TryGetProvider(WebHostDefaults.ServerUrlsKey, out IConfigurationProvider provider))
+                    {
+                        provider.Set(WebHostDefaults.ServerUrlsKey, aspnetUrls);
+                    }
+
+                    // The DOTNET_* and ASPNETCORE_* environment variables were added as part of the host configuration
+                    // phase. Before the phase is completed, add a configuration source that will conditionally block
+                    // reading of the Urls options so that they are not picked up by future Kestrel configuration callbacks.
+                    builder.Add(new ServerUrlsBlockingConfigurationSource(manager));
+                })
+                .ConfigureAppConfiguration((IConfigurationBuilder builder) =>
+                {
+                    // The settings.json, key-per-file, and DOTNETMONITOR_* environment variables were added as part
+                    // of the app configuration phase. Before the phase is completed, add a configuration source that will
+                    // conditionally block reading of the Urls options so that they are not picked up by future Kestrel
+                    // configuration callbacks.
+                    builder.Add(new ServerUrlsBlockingConfigurationSource(manager));
+                })
+                .ConfigureContainer((HostBuilderContext context, IServiceCollection services) =>
+                {
+                    // Container configuration is the last phase of building the host before the service provider is constructed.
+                    // At this point, all configuration callbacks have been executed. Lift the block on the Urls option so that
```

Comment:
Is this order something we can rely on long term? For example what if another phase is added in the future?

Reply:
I believe it is something we can fairly rely on. These phases haven't changed since .NET Core 2.1. All of the frameworks (runtime and aspnetcore) add all of their configuration sources during the host configuration phase. We add configuration sources to both host configuration and app configuration phases. Hence why the ServerUrlsBlockingConfigurationSource is added at the end of each of these phases. Since configurations sources cannot be added during service and container phases, don't think this will pop up unexpectedly. I've also added tests to help check if this occurs.